apf/apfH1Shapes.cc

```diff
@@ -0,0 +1,769 @@
+
+  apf::DynamicArray<apf::Vector3> nodes(non);
+  nodes.setSize(non);
+
+
+  int o = 0;
+  // vertices
+  nodes[o][0] = cp[0]; nodes[o][1] = cp[0]; nodes[o][2] = cp[0];
+  o++;
+  nodes[o][0] = cp[p]; nodes[o][1] = cp[0]; nodes[o][2] = cp[0];
+  o++;
+  nodes[o][0] = cp[0]; nodes[o][1] = cp[p]; nodes[o][2] = cp[0];
+  o++;
+  nodes[o][0] = cp[0]; nodes[o][1] = cp[0]; nodes[o][2] = cp[p];
+  o++;
+
+  // edges
+  for (int i = 1; i < p; i++) // (0,1)
+  {
+    nodes[o][0] = cp[i];  nodes[o][1] = cp[0];  nodes[o][2] = cp[0];
+    o++;
+  }
+
+  for (int i = 1; i < p; i++) // (1,2)
+  {
+    nodes[o][0] = cp[p-i];  nodes[o][1] = cp[i];  nodes[o][2] = cp[0];
+    o++;
+  }
+
+  for (int i = 1; i < p; i++) // (2,0)
+  {
+    nodes[o][0] = cp[0];  nodes[o][1] = cp[p-i];  nodes[o][2] = cp[0];
+    o++;
+  }
+
+  for (int i = 1; i < p; i++) // (0,3)
+  {
+    nodes[o][0] = cp[0];  nodes[o][1] = cp[0];  nodes[o][2] = cp[i];
+    o++;
+  }
+
+  for (int i = 1; i < p; i++) // (1,3)
+  {
+    nodes[o][0] = cp[p-i];  nodes[o][1] = cp[0];  nodes[o][2] = cp[i];
+    o++;
+  }
+
+  for (int i = 1; i < p; i++) // (2,3)
+  {
+    nodes[o][0] = cp[0];  nodes[o][1] = cp[p-i];  nodes[o][2] = cp[i];
+    o++;
+  }
+
+
+  // faces
+  // (0,1,2)
+  for (int j = 1; j < p; j++)
+    for (int i = 1; i + j < p; i++) {
+      double w = cp[i] + cp[j] + cp[p-i-j];
+      nodes[o][0] = cp[i]/w;  nodes[o][1] = cp[j]/w;  nodes[o][2] = cp[0];
+      o++;
+    }
+
+  // (0,1,3)
+  for (int j = 1; j < p; j++)
+    for (int i = 1; i + j < p; i++) {
+      double w = cp[i] + cp[j] + cp[p-i-j];
+      nodes[o][0] = cp[i]/w;  nodes[o][1] = cp[0]/w;  nodes[o][2] = cp[j]/w;
+      o++;
+    }
+
+  // (1,2,3)
+  for (int j = 1; j < p; j++)
+    for (int i = 1; i + j < p; i++) {
+      double w = cp[i] + cp[j] + cp[p-i-j];
+      nodes[o][0] = cp[p-i-j]/w;  nodes[o][1] = cp[i]/w;  nodes[o][2] = cp[j]/w;
+      o++;
+    }
+
+  // (0,2,3)
+  for (int j = 1; j < p; j++)
+    for (int i = 1; i + j < p; i++) {
+      double w = cp[i] + cp[j] + cp[p-i-j];
+      nodes[o][0] = cp[0];  nodes[o][1] = cp[i]/w;  nodes[o][2] = cp[j]/w;
+      o++;
+    }
+
+  // Region
+  for (int k = 1; k < p; k++) {
+    for (int j = 1; j + k < p; j++) {
+      for (int i = 1; i + j + k < p; i++) {
+        double w = cp[i] + cp[j] + cp[k] + cp[p-i-j-k];
+        nodes[o][0] = cp[i]/w;  nodes[o][1] = cp[j]/w;  nodes[o][2] = cp[k]/w;
+        o++;
+      }
+    }
+  }
+
+  // Populate T
+  mth::Matrix<double> T(non,non);
+  for (int m = 0; m < non; m++)
+  {
+    o = 0;
+    double x = nodes[m][0]; double y = nodes[m][1]; double z = nodes[m][2];
+
+    getChebyshevT(p, x, &shape_x[0]);
+    getChebyshevT(p, y, &shape_y[0]);
+    getChebyshevT(p, z, &shape_z[0]);
+    getChebyshevT(p, 1. - x - y - z, &shape_l[0]);
+
+    for (int k = 0; k <= p; k++)
+      for (int j = 0; j + k <= p; j++)
+        for (int i = 0; i + j + k <= p; i++)
+          T(o++, m) = shape_x[i]*shape_y[j]*shape_z[k]*shape_l[p-i-j-k];
+	}
```

Comment:
Is there a tab here?  The indentation looks odd.

Reply:
Fixed this.